core/network-peerset/src/lib.rs

```diff
@@ -0,0 +1,689 @@
+			expires: self.expires,
+			next_back_off: self.next_back_off,
+			back_off_until: self.back_off_until,
+			score: self.score.clone(),
+		}
+	}
+}
+
+#[derive(Debug, Clone)]
+struct AddrScore {
+	/// If connected, contains the moment when we connected. `None` if we're not connected.
+	connected_since: Option<Instant>,
+	/// Score of this address. Potentially needs to be updated based on `latest_score_update`.
+	score: u32,
+	/// When we last updated the score.
+	latest_score_update: Instant,
+}
+
+impl Addr {
+	/// Sets the addr to connected. If the score is lower than the given value, raises it to this
+	/// value.
+	fn connected_now(&mut self, raise_to_min: u32) {
+		let now = Instant::now();
+		Addr::flush(&mut self.score, now);
+		self.score.connected_since = Some(now);
+		if self.score.score < raise_to_min {
+			self.score.score = raise_to_min;
+		}
+	}
+
+	/// Applies a modification to the score.
+	fn adjust_score(&mut self, score_diff: i32) {
+		Addr::flush(&mut self.score, Instant::now());
+		if score_diff >= 0 {
+			self.score.score = cmp::min(MAX_SCORE, self.score.score + score_diff as u32);
+		} else {
+			self.score.score = self.score.score.saturating_sub(-score_diff as u32);
+		}
+	}
+
+	/// Sets the addr to disconnected and applies a modification to the score.
+	fn disconnected_now(&mut self, score_diff: i32) {
+		Addr::flush(&mut self.score, Instant::now());
+		self.score.connected_since = None;
+		if score_diff >= 0 {
+			self.score.score = cmp::min(MAX_SCORE, self.score.score + score_diff as u32);
+		} else {
+			self.score.score = self.score.score.saturating_sub(-score_diff as u32);
+		}
+	}
+
+	/// Returns true if we are connected to this addr.
+	fn is_connected(&self) -> bool {
+		self.score.connected_since.is_some()
+	}
+
+	/// Returns the score, and true if we are connected to this addr.
+	fn score_and_is_connected(&mut self) -> (u32, bool) {
+		Addr::flush(&mut self.score, Instant::now());
+		let is_connected = self.score.connected_since.is_some();
+		(self.score.score, is_connected)
+	}
+
+	/// Updates `score` and `latest_score_update`, and returns the score.
+	fn score(&mut self) -> u32 {
+		Addr::flush(&mut self.score, Instant::now());
+		self.score.score
+	}
+
+	fn flush(score: &mut AddrScore, now: Instant) {
+		if let Some(connected_since) = score.connected_since {
+			let potential_score: u32 = div_dur_with_dur(now - connected_since, CONNEC_DURATION_PER_SCORE);
+			// We flush when we connect to an address.
+			debug_assert!(score.latest_score_update >= connected_since);
+			let effective_score: u32 =
+				div_dur_with_dur(score.latest_score_update - connected_since, CONNEC_DURATION_PER_SCORE);
+			let to_add = potential_score.saturating_sub(effective_score);
+			score.score = cmp::min(MAX_SCORE, score.score + to_add);
+		}
+
+		score.latest_score_update = now;
+	}
+}
+
+/// Divides a `Duration` with a `Duration`. This exists in the stdlib but isn't stable yet.
+// TODO: remove this function once stable
```

Comment:
```suggestion
// TODO: remove this function once duration devision is stabilized - https://github.com/rust-lang/rust/issues/54361
```

Adding the link to the rust issue.